A reorganization of timekeeping to work with general calendars

xylar opened the pull request:
The Date class has been replaced by a few utility functions:
 * `stringToDatetime` - converts an MPAS date string to a `datetime.datetime`
 * `stringToRelativeDelta` - converts an MPAS date string to an `MpasRelativeDelta`
* `clampToNumpyDatetime64` - clamps a datetime to the valid range supported by xarray

A new `MpasRelativeDelta` class has been added (deriving from `dateutil.relativedelta.relativedelta`)

All analysis functions and tests have been updated to work with these utility functions.

xylar commented:
This branch needs to be rebased after #86 gets merged.

xylar commented:
# Testing

So far, I have tested only on my laptop using a G-240km test case (and skipping `regriddedSeaIceConcThick` because, as usual, I don't have the mapping file).

I plan to perform more thorough testing on Edison in the next few days.

xylar commented:
This branch implements the design in #103

pwolfram commented:
@xylar, I'll be happy to do a review following the rebase.

xylar commented:
@pwolfram, when you get around to reviewing this PR, please take a look at the CI I added for testing my new MpasRelativeDelta class near leap days. Feel free to suggest other tests I should add if you can think of any.

pwolfram commented on mpas_analysis/shared/timekeeping/MpasRelativeDelta.py:
```diff
@@ -0,0 +1,131 @@
+        return self.__neg__().__radd__(other)
+
+    def __sub__(self, other):
+        return NotImplemented
```
Is there a reason why subtraction isn't supported but addition is supported?  You also have negation, which suggests subtraction should work as a combination of addition and negation (or multiplication with a negative factor), e.g., A - B = A + (-B).  This should be commented or implemented.

xylar replied:
So that's because date - offset is sensible but offset - date doesn't really make sense.  So `__rsub__` is okay but `__sub__` should be undefined.  Have I missed something?  Keep in mind that addition and subtraction is with datetime.datetime objects, not with other `MpasRelativeDelta` objects.

pwolfram replied:
I would argue that the subtraction operation needs to be commutative so that offset - date should make sense because it is just `offset - date = - (date - offset)`.  This is because the negation operator is defined and I'm using standard arithmetic set operations, which I think make sense in this context and is presumed by many.  Alternatively, we should issue a warning that this is not implemented, which is true in general for operators like this that seem like they should intuitively be implemented.

xylar replied:
Negation operator on a date doesn't make sense and puts a date out of range.  So I don't think either `offset - date` or `- (date - offset)` are supported.  I don't believe `datetime.datetime` supports negation, and I think that is correct.

xylar replied:
So I would say that `offset1 - offset2 = - (offset2 - offset1)` if you're using two MpasRelativeDeltas but if you're using `datetime.datetime` as one of the objects, that's no longer necessarily a sensible operator.

pwolfram commented on mpas_analysis/shared/timekeeping/MpasRelativeDelta.py:
```diff
@@ -0,0 +1,131 @@
+    Only relative intervals (years, months, etc.) are supported and not the
+    absolute date specifications (year, month, etc.).  Addition/subtraction
+    of datetime.datetime objects (but not other MpasRelativeDelta,
+    datetime.timedelta or other related objects) is supported.
```
Why is this the case?  Additional comments are needed or this restriction should be removed.

xylar replied:
I don't think we have a use case where we need to be able to add or subtract two offsets.  I'm not sure if it's allowed for timedeltas.  I thought not but I'll check.

xylar replied:
Okay, it looks like timedelta supports addition and subtraction of other timedeltas so I'll implement it.  It doesn't seem that hard.

pwolfram replied:
We should test all overloaded operator operations too because this is best practices and will ensure that we get what is expected.

xylar replied:
I believe I'm testing them all at this point.

pwolfram commented on mpas_analysis/shared/timekeeping/utility.py:
```diff
@@ -0,0 +1,229 @@
+    """
+    Temporary function for adding an offset year and clamping a datetime to
+    range supported by `numpy.datetime64`.
+    """
```
Should this issue a warning (perhaps that can be turned off via a flag) that specifies which direction the clamp, i.e., `floor` or `ceil`, was performed?

xylar replied:
Why would you call this function if you didn't want the clamping feature?  What is the situation where you would want it clamped in one direction but not the other?

This function will go away in my next PR after this anyway, so I don't have much interest in adding functionality to it...

xylar replied:
I see, I had misread your comment the first time.  There are a lot of cases where this function is used precisely because the clamping behavior is desired, so it would be important to suppress the warning in those cases.  I'm inclined not to add the complexity required to have an optional warning, given that this function really is temporary and the next PR I am planning to make will eliminate this function entirely, along with yearOffset.  But I can add this functionality if you feel strongly about it.

pwolfram replied:
If we are going to get rid of this on a short timescale of weeks, but not months, this is noise and we should proceed in the less rigorous way.

xylar replied:
It's my intention to have this replaced as soon as possible (definitely weeks rather than months).  That partly depends on the speed of reviews and testing of the PRs that come before it.  So I appreciate you working on this.

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,200 @@
+        self.assertEqual(date1, date2)
+
+        # gregorian
+        calendar = 'gregorian'
```
How about looping over `gregorian` and `gregorian_noleap`?

xylar replied:
Yep, I'll do that.

pwolfram replied:
We want to test the whole range of acceptable dates for both and then the distinction between `gregorian` and `gregorian_noleap` for each distinguishing case, e.g., Feb 29.

xylar replied:
Really?  We can do that, but that seems pretty extreme to me.

xylar replied:
I guess I don't really understand what you mean by "test the whole range of acceptable dates"?  I guess you mean the different string formattings?  I think it's pretty redundant to check this for both calendars because the function for parsing these strings is completely independent of the calendar, it just keeps the calendar around for doing math operations later on.  So in any cases where we're not doing math, the calendar is irrelevant.

xylar replied:
But I'm looping over both calendars anyway.  So I guess this is resolved.

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,200 @@
+        date2 = datetime.datetime(year=2262, month=1, day=1)
+        self.assertEqual(date1, date2)
+
+        # test if the calendars behave as they should
```
The loop could go here and reduce the number of lines of code required by about a factor of two.

xylar replied:
I don't mind doing this and I agree it makes sense to systematically test both calendars but there will be no large reduction in code.  Each test is unique (except for the first 2) independent of calendar.

pwolfram replied:
Right, this makes sense and will clarify the code albeit not at a huge reduction in the amount of code.  It will be clearer that this set of cases responds the same for each calendar and this set responds differently.  Perhaps you could do something like I'll show you below in another comment... to make the distinction clearer.

xylar replied:
Okay, I think this is addressed.

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,200 @@
+                                      calendar='gregorian')
+        diff = date1-date2
+        self.assertEqual(diff, stringToDatetime('1995-12-26'))
+
```
I would like to see some tests on operators for `MpasRelativeDelta` instances.

xylar replied:
Many of these tests already do test addition and subtraction operators.  I will add some more systematic tests of all operators, though.

pwolfram replied:
This would be good. But you could also note that a particular test is doing both in a comment too if that is cleaner.

xylar replied:
Yep, I've done so.

pwolfram commented:
@xylar, I added some incomplete and preliminary comments on this PR to get you started.

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,200 @@
+        self.assertEqual(stringToDatetime('2016-03-01') -
+                         stringToRelativedelta('0000-00-01',
+                                               calendar='gregorian'),
+                         stringToDatetime('2016-02-29'))
```
How about simplifying via something like:

```python
for calendar, expected in zip(['gregorian', 'gregorian_noleap'], ['2016-03-01', '2016-02-29']):
    self.assertEqual(stringToDateTime('2016-03-01') - stringToRelativedelta('0000-00-01', calendar=calendar),
                                stringToDateTime(expected))
```
This more clearly shows the distinction between the calendars and their usage and reduces the amount of needed code.

xylar replied:
Okay, I'll do that.

xylar commented:
@pwolfram, I think I've addressed all your comments for now.  Why don't you wait until we can merge #86 and I can rebase before continuing your review?

xylar commented:
@pwolfram and @milenaveneziani, this PR has been rebased onto develop and is ready to review.  I have left on the "Don't Merge" flag because I want to squash some commits before it gets merged.

xylar commented:
Okay, I have tested this PR on Edison (while /project is still available!) with the usual beta0 run (in configs/edison) and results are bit-for-bit to those from `develop`.

I squashed commits, so the PR is now ready to merge.

@pwolfram, please let me know if there are remaining concerns that I haven't addressed.

@milenaveneziani, please test and merge once you get the go-ahead from @pwolfram.

pwolfram commented on mpas_analysis/shared/timekeeping/MpasRelativeDelta.py:
```diff
@@ -0,0 +1,154 @@
+    __truediv__ = __div__
+
+    def __repr__(self):
+        l = []
```
This is dangerous for multiple reasons-- I would greatly prefer that we never use single-character variables unless they are an index.  Furthermore, "l" looks a lot like a "1", especially in certain fonts.  Cna you please make this something like `rep` or `strrep`, e.g. something three characters or more instead of `l`?

xylar replied:
Sorry, I copied this code straight out of `relativedelta` and only modified the minimum to add the calendar.  I agree that it's badly written.  I'll clean it up.

pwolfram commented on mpas_analysis/shared/timekeeping/utility.py:
```diff
@@ -0,0 +1,229 @@
+    Last modified
+    -------------
+    02/04/2017
+    """
```
An aside-- we should set something up that produces a "readthedocs" descrition of at least our public API.  Feel free to do the honors for an issue or I can also do so.  I'm happy to set this up when I get a a moment too if you would like so that we can leverage all your really good work getting docstrings into a more general sphinx-compatable format.

xylar replied:
Go for it.  I agree that this needs to happen but I haven't had time (and won't for awhile) to look into it further.

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,231 @@
+        # YYYY-MM-DD
+        # SSSSS
+
+        for calendar in ['gregorian', 'gregorian_noleap']:
```
Great!  Thanks @xylar :-)

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,231 @@
+        date2 = datetime.datetime(year=2262, month=1, day=1)
+        self.assertEqual(date1, date2)
+
+    def test_MpasRelativeDeltaOps(self):
```
Great!

pwolfram commented on mpas_analysis/test/test_timekeeping.py:
```diff
@@ -0,0 +1,231 @@
+
+        # make sure there's an error when we try to add MpasRelativeDeltas
+        # with different calendars
+        with self.assertRaisesRegexp(ValueError,
```
Perfect, because this enforces a design intent.

xylar commented:
@pwolfram, I think I've addressed your remaining comment.

pwolfram commented:
@xylar, I'm giving the go-ahead to @milenaveneziani for the merge.

milenaveneziani commented:
I tested this on wolf using the GMPAS-240 case and including the sea-ice climos. All went well.
Here is the sea-ice concentration for JFM, for example:
![iceconcnasateamnh_20170120 beta0 gmpas-qu240 wolf_jfm_years0002-0003](https://cloud.githubusercontent.com/assets/8237579/22835938/c062e1a6-ef77-11e6-909d-e675245e121c.png)
